src/Arrays/AppendedArrays.jl

```diff
@@ -63,7 +63,8 @@ struct AppendedArray{T,A,B} <: AbstractVector{T}
     B = typeof(b)
     ai = testitem(a)
     bi = testitem(b)
-    T = eltype(collect((ai,bi)))
+    #T = eltype(collect((ai,bi)))
+    T = promote_type(typeof(ai),typeof(bi))
```

Comment:
Using `promote_type` breaks the tests. What is the action to be taken here ? Any other options ?

Reply:
Ok. I think I understand why the test fails (https://github.com/gridap/Gridap.jl/runs/5178272647?check_suite_focus=true#step:6:163). 

In the following function 

```
function Base.getindex(v::AppendedArray,i::Integer)
  l = length(v.a)
  if i > l
    v.b[i-l]
  else
    v.a[i]
  end
end
```

we are not performing the type conversion when either the eltype of v.b or v.a is different from the one of v, and this breaks one postcondition of getindex, namely that the type of the returned item has to be equivalent to the eltype of the array on which you invoke getindex. The same should be done in the implementation of getindex!. BTW, this is the type conversion I was referring to here https://github.com/gridap/Gridap.jl/pull/653#discussion_r804447773, although I though it was implicit, it has to be done explicitly.

I think this will fix the breaking tests. Another question is whether we can afford/we want afford this. I think there is price to pay if we mix arrays of different types, we cannot solve that.

Reply:
How important is it for the `AppendedArray` to be a subtype of `AbstractVector{T}`? If the appended array was its own object which was built from 2 concrete type arrays `AppendedArray{T1,T2}` would it alleviate the issue? In that case perhaps you dont need to try and convert on type to the other.

Reply:
It is essential. 

Otherwise, we cannot leverage the software which is written in terms of the abstract interface of `AbstractVector{T}`.

We have to live with this constraint. Living with this constraint implies overhead, either in the form of the explicit type conversions or type instability. 

Other alternative is to perform the conversion before setting up the AppendedArray in the calling code, and do not let the latter manage this scenario.(I think you mentioned that at some point).

Reply:
Ok i see. Which solution do you suggest? I will have a go at implementing it.

Reply:
I would go for this option 

> Other alternative is to perform the conversion before setting up the AppendedArray in the calling code, and do not let the latter manage this scenario.(I think you mentioned that at some point).

as far as there is no other showstopper that I am missing.

Reply:
ok i made a commit. The problem is now that it breaks a different test because performing the conversion early changes the objects from a `Fill` :
```julia
cellmat = integrate( ∇(dv)⋅∇(du), quad )
@test isa(cellmat.a,Fill)
```
Should I try instead to do the explicit conversions after the creation of the `AppendedArray` instead?

Reply:
Wait. This is not why I was thinking of. I was thinking to leave the `AppendedArray` constructor untouched (as it is in the main trunk in Gridap). The code that calls the constructor is the one that has to convert the input arrays. Do you know what I mean?

Reply:
Again, I do not have the whole picture of the calling code in mind. There might be some showstopper that I am not able to anticipate.

Reply:
ok i made another attempt here. Is this last commit more like what you mean ?

Reply:
Ok. I have now realized about something that I was not aware (my bad, apologies). Let me explain. 

Before any modification in this PR, the constructor of `AppendedArray` was ALREADY defined as:

```  
function AppendedArray(a::AbstractArray,b::AbstractArray)
    A = typeof(a)
    B = typeof(b)
    ai = testitem(a)
    bi = testitem(b)
    T = eltype(collect((ai,bi)))
    new{T,A,B}(a,b)
  end
```

This is indeed how you have left it in the last commit (as per my request) [up to the fact you have changed the order of the instructions, without affecting the result, though.] 

This implementation of the constructor already handles the case in which the eltype of `a` and the eltype of `b` are different. (Indeed, I have seen that there are some tests of AppendedArrays in the test suite that check that everything ok in this particular scenario). 

I though (wrongly) that the implementation of the `AppendedArray`constructor was such it was NOT able to handle that particular case, and that `T = eltype(collect((ai,bi)))` was actually your addition. Indeed, I though (wrongly) that the constructor (before any modification) was imposing the eltype of the two arrays to be equivalent, and assert otherwise. 

It turns out that `eltype(collect((ai,bi)))` is equivalent to `eltype((ai,bi))`. I have been taking at the implementation of this latter function here https://github.com/JuliaLang/julia/blob/master/base/tuple.jl#L146. While I do not get all details, it is clear that `eltype((ai,bi))` can return a polymorphic (abstract) eltype if the types of ai and bi are different. For example, in this simple example:

```
julia> eltype((1,3.0))
Real

julia> Real.abstract
true
```

And this seems what @fverdugo wanted to avoid in his initial comment here: https://github.com/gridap/Gridap.jl/pull/653#discussion_r705005893 

Thus, now, I am little bit lost/confused. @fverdugo can you please help to clarify the situation?

Reply:
ok, I now that you @ConnorMallon  already mentioned something in this direction here https://github.com/gridap/Gridap.jl/pull/653#issuecomment-924514968